processor/ratelimitprocessor/gubernator.go

```diff
@@ -116,19 +149,50 @@ func (r *gubernatorRateLimiter) Shutdown(context.Context) error {
 		r.clientConn = nil
 	}
 	r.client = nil
+	if c, ok := r.classResolver.(component.Component); ok {
+		if err := c.Shutdown(ctx); err != nil {
+			return fmt.Errorf("failed to shutdown class resolver: %w", err)
+		}
+	}
 	return nil
 }
 
 func (r *gubernatorRateLimiter) RateLimit(ctx context.Context, hits int) error {
 	uniqueKey := getUniqueKey(ctx, r.cfg.MetadataKeys)
-	cfg := resolveRateLimitSettings(r.cfg, uniqueKey)
-	now := time.Now()
-
+	// First resolve the class if classes are set.
+	class, err := r.classResolver.ResolveClass(ctx, uniqueKey)
+	if err != nil {
+		r.telemetryBuilder.RatelimitResolverFailures.Add(ctx, 1,
+			metric.WithAttributeSet(
+				attribute.NewSet(attribute.String("unique_key", uniqueKey)),
+			),
+		)
+		r.logger.Warn("class resolver failed, falling back",
+			zap.Error(err),
+			zap.String("unique_key", uniqueKey),
+			zap.String("default_class", r.cfg.DefaultClass),
+		)
+	}
+	// Resolve rate limit precedence:
+	// override -> class -> default_class -> fallback.
+	cfg, sourceKind, className := resolveRateLimit(r.cfg, uniqueKey, class)
 	rate, burst := cfg.Rate, cfg.Burst
+	now := time.Now()
+	// If dynamic rate limiting is enabled and not disabled for this request,
+	// calculate the dynamic rate and burst.
 	if r.cfg.DynamicRateLimiting.Enabled && !cfg.disableDynamic {
+		attrs := metric.WithAttributeSet(attribute.NewSet(
+			attribute.String("source_kind", string(sourceKind)),
+			attribute.String("class", className),
+		))
 		rate, burst = r.calculateRateAndBurst(ctx, cfg, uniqueKey, hits, now)
-		if rate < 0 {
-			return fmt.Errorf("error calculating dynamic rate limit for unique key %s", uniqueKey)
+		if rate < 0 { // Degraded mode - Gubernator unreachable. Fallback to static rate.
+			r.telemetryBuilder.RatelimitGubernatorDegraded.Add(ctx, 1, attrs)
```

Comment:
nit: Can we have a single metric event with single dimension add these dynamic/static attributes instead of having them separate?

Reply:
I'll look into that.

Reply:
Fixed